src/lib.rs

```diff
@@ -195,11 +195,42 @@ extern crate smallvec;
 pub type FxHashBuilder = hashbrown::hash_map::DefaultHashBuilder;
 
 use std::collections::hash_map::RandomState;
+use std::fmt;
 use std::hash::{BuildHasher, Hash};
+use std::sync::Arc;
 
 mod inner;
 use inner::Inner;
 
+/// Unary predicate used to retain elements
+#[derive(Clone)]
+pub struct Predicate<V>(pub(crate) Arc<Fn(&V) -> bool + Send + Sync>);
```

Comment:
I _think_ it should be possible to make this a `Box` instead of an `Arc`?

Reply:
I had it as a `Box` earlier, but it required removing `Eq` and `Clone` from the `Operation` enum. It doesn't add any overhead to calling the function, but it is cloneable.

Reply:
Hmm, yeah, I suppose we sort of want to use `shallow_copy` for this `Fn` too, but that's trickier to achieve. `Arc` is probably fine.

Reply:
Somehow I didn't see this. 

Shallow copy isn't needed for this at all. This doesn't carry any value nor is it inserted into the map itself. It's simply a callback used to evaluate the `retain` operation.